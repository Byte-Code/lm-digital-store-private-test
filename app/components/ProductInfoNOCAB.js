import React, { Component } from 'react';
import ImmutablePropTypes from 'react-immutable-proptypes';
import styled from 'styled-components';

const Wrapper = styled.div`
  width: 100%;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  background-color: #efefef;
  padding: 69px 40px 57px;
  &>p {
    font-size: 18px;
    line-height: 28px;
    width: 425px;
    font-family: LeroyMerlinSans Light;
    width: 50%;
    &>span {
      font-family: LeroyMerlinSans;
    }
  }
`;

const Divider = styled.div`
  border: 1px dashed #67cb33;
  width: 100%;
  margin-top: 20px;
`;

export default class ProductInfoNOCAB extends Component {
  static propTypes = {
    descriptions: ImmutablePropTypes.list.isRequired
  }

  renderDescriptions() {
    const { descriptions } = this.props;
    if (!descriptions) {
      return null;
    }
    return descriptions.get(0).map(item => (
      <p key={`${item.get('label')}${item.get('value')}`}>
        <span>{item.get('label') && `${item.get('label')}: `}</span>
        {item.get('value')}
      </p>
    ));
  }

  render() {
    return (
      <Wrapper>
        {this.renderDescriptions()}
        <Divider />
      </Wrapper>
    );
  }
}
