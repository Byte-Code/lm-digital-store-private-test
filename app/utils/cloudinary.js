import cloudinary from 'cloudinary-core';

const cloudinaryName = 'leroymerlin';

const cl = cloudinary.Cloudinary.new({ cloud_name: cloudinaryName });
cl.config('cdn_subdomain', true);

export default function getUrl(imageId, options = {}) {
  const cleanImageId =
      imageId && imageId.charAt(0) === '/'
          ? imageId.substr(1)
          : imageId;

  let url = cl.url(cleanImageId, {
    crop: 'pad',
    default_image: 'no_image_available',
    flags: 'progressive',
    ...options,
  });

  if (url) url = url.replace(/^(https?):\/\//, '//');

  return url;
}
