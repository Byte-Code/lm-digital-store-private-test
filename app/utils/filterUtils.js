import { fromJS, List, Set } from 'immutable';

// GENERAL
export function isValidList(list) {
  return list && !list.isEmpty();
}

export function buildFilterMap(query, categoryCode) {
  return fromJS({
    activeAid: buildAid(query),
    activeFilters: buildFilters(query),
    activeAvailability: buildAvailability(query),
    categoryCode
  });
}

export function filterCatalogue(idsByAids, idsByFilters, idsByAvailability) {
  let toIntersect = fromJS([idsByAvailability]);
  if (idsByAids) {
    toIntersect = toIntersect.push(idsByAids);
  }
  if (idsByFilters) {
    toIntersect = toIntersect.push(idsByFilters);
  }
  return Set.intersect(toIntersect.toJS());
}

// AIDS
export function buildAid(query) {
  const aids = query.aids ? decodeURIComponent(query.aids) : '';
  return aids;
}

export function filterProductsByAid(sellingAids, activeAid) {
  if (isValidList(sellingAids) && activeAid) {
    return sellingAids.find(a => a.get('code') === activeAid).get('products').toSet();
  }
  return null;
}

// FILTERS
export function buildFilters(query) {
  if (query.filters) {
    return fromJS(decodeURIComponent(query.filters).split('&'));
  }
  return List();
}

export function getProductsByFilter(filterGroups, activeFilters) {
  return filterGroups
    .map(g =>
      g
        .get('filters')
        .reduce((acc, f) => {
          if (activeFilters.includes(f.get('code'))) {
            return acc.push(f.get('products'));
          }
          return acc;
        }, List())
        .flatten()
        .toSet()
    )
    .filterNot(f => f.isEmpty())
    .toArray();
}

export function filterProducts(filterGroups, activeFilters) {
  if (isValidList(filterGroups) && isValidList(activeFilters)) {
    return Set.intersect(getProductsByFilter(filterGroups, activeFilters));
  }
  return null;
}

// AVAILABILITY
export function buildAvailability(query) {
  return query.availability === 'true';
}

export function filterProductsByAvailability(productList, activeAvailability) {
  if (isValidList(productList)) {
    let products = productList;
    if (activeAvailability) {
      products = productList.filter(p => p.get('storeStock') > 0);
    }
    return products.map(p => p.get('code')).toSet();
  }
  return Set();
}
