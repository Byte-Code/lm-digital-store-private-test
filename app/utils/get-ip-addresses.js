const os = require('os');
const interfaces = os.networkInterfaces();

export default function getIpAddresses() {
  const addresses = [];
  for (let k in interfaces) {
    for (let k2 in interfaces[k]) {
      const address = interfaces[k][k2];
      if (address.family === 'IPv4' && !address.internal) {
        addresses.push(address.address);
      }
    }
  }
  
  return addresses;
}
