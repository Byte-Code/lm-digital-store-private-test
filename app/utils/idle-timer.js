class IdleTimer {
  constructor() {
    this.idleTime = 0;
    this.tresholdReached = false;
    this.handleEvent = this.handleEvent.bind(this);
    this.tick = this.tick.bind(this);
  }

  init(time, onStart, onComplete, treshold, onReachTreshold, onReset) {
    this.time = time;
    this.onStart = onStart;
    this.onComplete = onComplete;
    this.onReachTreshold = onReachTreshold;
    this.treshold = treshold;
    this.onReset = onReset;
    document.addEventListener('pointerenter', this.handleEvent, false);
    document.addEventListener('pointermove', this.handleEvent, false);
    document.addEventListener('scroll', this.handleEvent, false);
    this.start();
  }

  handleEvent() {
    this.resetIndleTime();
  }

  tick() {
    this.idleTime = this.idleTime + 1;
    if (!this.tresholdReached && this.idleTime >= this.treshold / 1000) {
      this.tresholdReached = true;
      this.onReachTreshold();
    }
    if (this.idleTime >= this.time / 1000) {
      if (this.onComplete) {
        this.onComplete();
      }
      this.resetIndleTime();
    }
  }

  start() {
    clearInterval(this.interval);
    if (this.onStart) {
      this.onStart();
    }
    this.interval = setInterval(this.tick, 1000);
  }

  resetIndleTime() {
    if (this.onReset) {
      this.onReset();
    }
    this.tresholdReached = false;
    this.idleTime = 0;
  }
}

export default new IdleTimer();
