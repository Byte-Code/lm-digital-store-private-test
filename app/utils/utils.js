import { Range, List, Set } from 'immutable';

export function titleFormatter(text) {
  return text.charAt(0) + text.slice(1).toLowerCase();
}

export function chunkItemList(itemList, chunkSize) {
  return Range(0, itemList.count(), chunkSize)
    .map(chunkStart => itemList.slice(chunkStart, chunkStart + chunkSize));
}

// AIDS
export function buildAid(query) {
  const aids = query.aids ? decodeURIComponent(query.aids) : '';
  return aids;
}

export function filterProductsByAid(sellingAids, activeAid) {
  if (sellingAids.isEmpty()) {
    return Set();
  }
  if (!activeAid) {
    return sellingAids.reduce((acc, val) => acc.push(val.get('products')), List()).flatten().toSet();
  }
  return sellingAids.find(a => a.get('code') === activeAid).get('products').toSet();
}

// FILTERS
export function buildFilters(query) {
  const filters = query.filters ? List(query.filters.split(',')) : List();
  return filters.map(facet => decodeURIComponent(facet));
}

function getAllFilterProducts(filterGroups) {
  return filterGroups.map(g => (g.get('filters')
    .reduce((acc, f) => acc.push(f.get('products')), List()))
  ).flatten().toSet();
}

function getProductsByFilter(filterGroups, activeFilters) {
  return filterGroups.map(g =>
    g.get('filters').reduce((acc, f) => {
      if (activeFilters.includes(f.get('code'))) {
        return acc.push(f.get('products'));
      }
      return acc;
    }, List()).flatten().toSet()
  ).filterNot(f => f.isEmpty()).toArray();
}

export function filterProducts(filterGroups, activeFilters) {
  if (activeFilters.isEmpty()) {
    return getAllFilterProducts(filterGroups);
  }
  return Set.intersect(getProductsByFilter(filterGroups, activeFilters));
}
