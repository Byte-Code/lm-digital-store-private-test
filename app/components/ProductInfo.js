import React, { Component } from 'react';
import ImmutablePropTypes from 'react-immutable-proptypes';
import styled from 'styled-components';
import AddIcon from 'material-ui/svg-icons/content/add-circle-outline';

import { titleFormatter } from '../utils/stringUtils';
import Card from './Card';

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
`;

const Row = styled.div`
  width: 100%;
  display: flex;
  flex-wrap: wrap;
  background-color: ${props => props.bgColor}
  padding-top: ${props => props.pTop};
  padding-left: 125px;
  padding-right: 125px;
  padding-bottom: 57px;
  &>div:first-child {
    margin-right: 100px;
  }
`;

const Column = styled.div`
  width: calc((100% - 100px) /2);
  &>div:first-child {
    margin-bottom: 30px;
  }
`;

const Section = styled.div`
  &>p{
    font-size: 14px;
    margin-bottom: 17px;
    font-family: LeroyMerlinSans Light;
    &>span{
      font-family: LeroyMerlinSans;
    }
  }
`;

const CardTitle = styled.div`
  display: flex;
  align-items: center;
  padding: 19px 125px;
  background: #f7f7f7;
  &>h3{
    margin-left: 20px;
    font-size: 16px;
    line-height: 32px;
  }
`;

const Title = styled.div`
  font-family: LeroyMerlinSans Italic;
  font-size: 16px;
  line-height: 20px;
  color: #339900;
  margin-bottom: 14px;
`;

const Divider = styled.div`
  border: 1px dashed #67cb33;
  width: 100%;
  margin-top: 18px;
`;

function CardTitleComponent() {
  return (
    <CardTitle>
      <AddIcon color="#333333" />
      <h3>Maggiori informazioni</h3>
    </CardTitle>
  );
}

export default class ProductInfo extends Component {
  static propTypes = {
    marketingDescriptions: ImmutablePropTypes.map.isRequired,
    descriptions: ImmutablePropTypes.list.isRequired
  }

  renderBenefits() {
    const { marketingDescriptions } = this.props;
    const benefits = marketingDescriptions.get('benefits');

    if (!benefits) {
      return null;
    }

    return (
      <Section>
        <Title>Vantaggi</Title>
        {benefits.map(b => <p key={b}>{b}</p>)}
        <Divider />
      </Section>
    );
  }

  renderBlocks() {
    const { marketingDescriptions } = this.props;
    const blocks = marketingDescriptions.get('chooseBlocks').reverse();

    return blocks.map(block => (
      <Section key={block.get('title')}>
        <Title>{titleFormatter(block.get('title'))}</Title>
        {block.get('customerChooses').map(c =>
          <p key={c.get('description')}>{c.get('description')}</p>)}
        {!block.equals(blocks.last()) && <Divider />}
      </Section>
    ));
  }

  renderDescriptions(code) {
    const { descriptions } = this.props;
    const category = descriptions.find(d => d.get('code') === code);
    const categoryTitle = category.get('title');
    const categoryItems = category.get('description');

    // should I map all of them?
    return (
      <Section>
        <Title>{titleFormatter(categoryTitle)}</Title>
        {categoryItems.map(item => (
          <p key={`${item.get('label')}${item.get('value')}`}>
            <span>{item.get('label') && `${item.get('label')}: `}</span>
            {item.get('value')}
          </p>
        ))}
        {category.get('code') === 'DIMENSIONI' && <Divider />}
      </Section>
    );
  }

  render() {
    return (
      <Wrapper>
        <Row pTop="69px" bgColor="#efefef">
          <Column>{this.renderBenefits()}</Column>
          <Column>{this.renderBlocks()}</Column>
        </Row>
        <Card TitleComponent={CardTitleComponent}>
          <Row pTop="31px" bgColor="#f7f7f7">
            <Column>{this.renderDescriptions('MATERIALE')}</Column>
            <Column>
              {this.renderDescriptions('DIMENSIONI')}
              {this.renderDescriptions('UTILE_DA_SAPERE')}
            </Column>
          </Row>
        </Card>
      </Wrapper>
    );
  }
}
