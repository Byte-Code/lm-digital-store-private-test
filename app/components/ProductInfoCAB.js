import React, { Component, PropTypes } from 'react';
import ImmutablePropTypes from 'react-immutable-proptypes';
import styled from 'styled-components';
import { List } from 'immutable';
import AddIcon from 'material-ui/svg-icons/content/add-circle-outline';
import RemoveIcon from 'material-ui/svg-icons/content/remove-circle-outline';

import { titleFormatter } from '../utils/utils';
import Card from './Card';

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
`;

const Row = styled.div`
  width: 100%;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  background-color: ${props => props.bgColor}
  padding-top: ${props => props.pTop};
  padding-left: 125px;
  padding-right: 125px;
  padding-bottom: 57px;
`;

const Column = styled.div`
  width: calc((100% - 100px) /2);
`;

const Section = styled.div`
  margin-bottom: 20px;
  &>p{
    font-size: ${props => props.fSize};
    font-family: LeroyMerlinSans Light;
    &>span{
      font-family: LeroyMerlinSans;
    }
  }
`;

const CardTitle = styled.div`
  display: flex;
  cursor: pointer;
  align-items: center;
  padding: 19px 125px;
  background: #f7f7f7;
  &>h3{
    margin-left: 20px;
    font-size: 16px;
    line-height: 32px;
  }
`;

export const Title = styled.div`
  font-family: LeroyMerlinSans Italic;
  font-size: ${props => props.fSize};
  line-height: 20px;
  color: #339900;
  margin-bottom: 14px;
`;

const Divider = styled.div`
  border: 1px dashed #67cb33;
  width: 100%;
  margin-top: 20px;
`;

function CardTitleComponent({ expanded }) {
  const Icon = expanded ? RemoveIcon : AddIcon;
  return (
    <CardTitle>
      <Icon color="#333333" />
      <h3>Maggiori informazioni</h3>
    </CardTitle>
  );
}

CardTitleComponent.propTypes = {
  expanded: PropTypes.bool.isRequired
};

export default class ProductInfoCAB extends Component {
  static propTypes = {
    marketingDescriptions: ImmutablePropTypes.map.isRequired,
    descriptions: ImmutablePropTypes.seq.isRequired
  }

  renderBlocks() {
    const { marketingDescriptions } = this.props;
    if (!marketingDescriptions) {
      return null;
    }
    // TODO why can this be undefined?
    const blocks = marketingDescriptions.get('chooseBlocks') || List();

    return blocks.map(block => (
      <Column key={block.get('title')}>
        <Section fSize="18px">
          <Title fSize="20px">{titleFormatter(block.get('title'))}</Title>
          {block.get('customerChooses').map(c =>
            <p key={c.get('description')}>{c.get('description')}</p>)}
        </Section>
        <Divider />
      </Column>
    ));
  }

  renderDescriptions(index) {
    const { descriptions } = this.props;
    const currentBatch = descriptions.get(index);
    if (!currentBatch) {
      return null;
    }

    return currentBatch.map(desc => (
      <Section fSize="14px" key={desc.get('code')}>
        <Title fSize="16px">{titleFormatter(desc.get('title'))}</Title>
        {desc.get('description').map(item => (
          <p key={`${item.get('label')}${item.get('value')}`}>
            <span>{item.get('label') && `${item.get('label')}: `}</span>
            {item.get('value')}
          </p>
        ))}
        {(descriptions.get(index).last() !== desc) && <Divider />}
      </Section>
    ));
  }

  render() {
    return (
      <Wrapper>
        <Row pTop="69px" bgColor="#efefef">
          {this.renderBlocks()}
        </Row>
        <Card TitleComponent={CardTitleComponent}>
          <Row pTop="31px" bgColor="#f7f7f7">
            <Column>
              {this.renderDescriptions(0)}
            </Column>
            <Column>
              {this.renderDescriptions(1)}
            </Column>
          </Row>
        </Card>
      </Wrapper>
    );
  }
}
