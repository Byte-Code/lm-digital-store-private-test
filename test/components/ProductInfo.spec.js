import React from 'react';
import { shallow } from 'enzyme';
import { fromJS, Map } from 'immutable';

import { chunkItemList } from '../../app/utils/utils';
import ProductInfo from '../../app/components/ProductInfo';
import ProductInfoCAB from '../../app/components/ProductInfoCAB';
import ProductInfoNOCAB from '../../app/components/ProductInfoNOCAB';

const typeCAB = 'CAB';
const typeNOCAB = 'foo';
const descriptions = fromJS(['descriptions0', 'descriptions1', 'descriptions2']);
const marketingDescriptions = Map();

describe('ProductInfo', () => {
  it('should render ProductInfoNOCAB with the right props if product is NOCAB', () => {
    const result = shallow(
      <ProductInfo
        productType={typeNOCAB}
        descriptions={descriptions}
        marketingDescriptions={marketingDescriptions}
      />
    );
    const nextProps = { descriptions };
    expect(result.find(ProductInfoCAB)).toHaveLength(0);
    expect(result.find(ProductInfoNOCAB)).toHaveLength(1);
    expect(result.find(ProductInfoNOCAB).props()).toEqual(
      nextProps
    );
    expect(result).toMatchSnapshot();
  });

  it('should render ProductInfoCAB with the right props if product is CAB', () => {
    const result = shallow(
      <ProductInfo
        productType={typeCAB}
        descriptions={descriptions}
        marketingDescriptions={marketingDescriptions}
      />
    );
    const halfDescriptionsSize = Math.ceil(descriptions.size / 2);
    const chunkedDescriptions = chunkItemList(descriptions, halfDescriptionsSize);
    const nextProps = { descriptions: chunkedDescriptions, marketingDescriptions };
    expect(result.find(ProductInfoNOCAB)).toHaveLength(0);
    expect(result.find(ProductInfoCAB)).toHaveLength(1);
    expect(result.find(ProductInfoCAB).props()).toEqual(
      nextProps
    );
    expect(result).toMatchSnapshot();
  });
});
