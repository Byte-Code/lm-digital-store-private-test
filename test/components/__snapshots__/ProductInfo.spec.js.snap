// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductInfo should render ProductInfoCAB with the right props if product is CAB 1`] = `
<ProductInfoCAB
  descriptions={
    Array [
      Array [
        "descriptions0",
        "descriptions1",
      ],
      Array [
        "descriptions2",
      ],
    ]
  }
  marketingDescriptions={Object {}}
/>
`;

exports[`ProductInfo should render ProductInfoNOCAB with the right props if product is NOCAB 1`] = `
<ProductInfoNOCAB
  descriptions={
    Array [
      "descriptions0",
      "descriptions1",
      "descriptions2",
    ]
  }
/>
`;
